osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,201 @@
+            # TODO: narrow down possible exceptions here
+            logging.error(e.message)
+        else:
+            if not resp.status == 201:
+                return None
+            json_resp = yield from resp.json()
+            return json_resp['data']['attributes']['token_id']
+
+        return
+
+    @asyncio.coroutine
+    def _find_or_create_user(self, username, password):
+        """ Checks to see if username exists in DB and compares password hashes.
+            Tries to authenticate and create user if a valid user/oauth_token is not found.
+
+        :return: user or None
+        """
+        user = None
+
+        try:
+            user = session.query(User).filter(User.osf_login == username).one()
+        except MultipleResultsFound:
+            logging.warning('Multiple users with same username. deleting all users with this username. restarting function.')
+            for user in session.query(User).filter(User.osf_login == username).all():
+                session.delete(user)
+                try:
+                    save(session, user)
+                except SQLAlchemyError as e:
+                    self.display_error('Unable to save user data. Please try again later.', exc=e.message)
+            user = yield from self._find_or_create_user(username, password)
+        except NoResultFound:
+            logging.debug('User doesnt exist. Attempting to authenticate, then creating user.')
+            personal_access_token = yield from self._authenticate(username, password)
+            if not personal_access_token:
+                self.failed_login = True
+                return
+            else:
+                user = User(
+                    full_name='',
+                    osf_id='',
+                    osf_login=username,
+                    osf_password=generate_hash(password),
+                    osf_local_folder_path='',
+                    oauth_token=personal_access_token,
+                )
+        finally:
+            #Hit APIv2 and populate more user data if possible
+            if not user:
+                self.display_error('Invalid login credentials')
+                return
+
+            if not user.oauth_token or user.osf_password != generate_hash(password, user.osf_password):
+                logging.warning('Login error: User {} either has no oauth token or submitted a different password. Attempting to authenticate'.format(user))
+                personal_access_token = yield from self._authenticate(username, password)
+                if not personal_access_token:
+                    self.failed_login = True
+                else:
+                    user.oauth_token = personal_access_token
+
+            if not self.failed_login:
+                return (yield from self._populate_user_data(user, username, password))
+            else:
+                return
+
+    @asyncio.coroutine
+    def _populate_user_data(self, user, username, password):
```

Comment:
Is it an intentional decision to not save the user object (created here) until line 175 of `log_in`?

(Asking because we had a different bug today where the lack of save was unintentional)